cpp/src/arrow/util/bitmap_ops.cc

```diff
@@ -298,52 +297,52 @@ void TransferBitmap(const uint8_t* data, int64_t offset, int64_t length,
     auto nwords = reader.words();
     while (nwords--) {
       auto word = reader.NextWord();
-      writer.PutNextWord(invert_bits ? ~word : word);
+      writer.PutNextWord(mode == TransferMode::Invert ? ~word : word);
     }
     auto nbytes = reader.trailing_bytes();
     while (nbytes--) {
       int valid_bits;
       auto byte = reader.NextTrailingByte(valid_bits);
-      writer.PutNextTrailingByte(invert_bits ? ~byte : byte, valid_bits);
+      writer.PutNextTrailingByte(mode == TransferMode::Invert ? ~byte : byte, valid_bits);
     }
-  } else {
-    // Shift dest by its byte offset
-    dest += dest_byte_offset;
+  } else if (length) {
+    int64_t num_bytes = BitUtil::BytesForBits(length);
+
+    // Shift by its byte offset
+    data += offset / 8;
+    dest += dest_offset / 8;
 
     // Take care of the trailing bits in the last byte
+    // E.g., if trailing_bits = 5, last byte should be
+    // - low  3 bits: new bits from last byte of data buffer
+    // - high 5 bits: old bits from last byte of dest buffer
     int64_t trailing_bits = num_bytes * 8 - length;
-    uint8_t trail = 0;
-    if (trailing_bits && restore_trailing_bits) {
-      trail = dest[num_bytes - 1];
-    }
+    uint8_t kTrailingBitmask = (1U << (8 - trailing_bits)) - 1;
```

Comment:
I meant use the `kTrailingBitmask` constant array that's exposed in `bit_util.h` :-)

Reply:
Ah, it's a lookup table, cool. Will change.

Reply:
We need a table {255, 127, 63, 31, 15, 7, 3, 1}, looks there's no such lookup table available.
Shall I add a new table or keep the code as is (revert the naming)?
https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bit_util.h#L420

Reply:
Hmm, let's simply revert the renaming then :-)

Reply:
done